src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -2014,4 +2138,594 @@ void CorProfiler::GetAssemblyAndSymbolsBytes(BYTE** pAssemblyArray, int* assembl
+  GetWrapperMethodRef(module_metadata, module_id, *method_replacement, wrapper_method_ref, wrapper_type_ref);
+
+  Debug("*** CallTarget_RewriterCallback() Start: ", caller->type.name, ".", caller->name, 
+       "() [IsVoid=", isVoid, 
+       ", IsStatic=", isStatic, 
+       ", IntegrationType=", method_replacement->wrapper_method.type_name,
+       ", Arguments=", numArgs, 
+       "]");
+
+  // *** Create rewriter
+  ILRewriter rewriter(this->info_, methodHandler->GetFunctionControl(), module_id, function_token);
+  bool modified = false;
+  auto hr = rewriter.Import();
+  if (FAILED(hr)) {
+    Warn("*** CallTarget_RewriterCallback(): Call to ILRewriter.Import() failed for ", module_id, " ", function_token);
+    return hr;
+  }
+
+  // *** Store the original il code text if the dump_il option is enabled.
+  std::string original_code;
+  if (dump_il_rewrite_enabled) {
+    original_code = GetILCodes(
+        "*** CallTarget_RewriterCallback(): Original Code: ", &rewriter,
+        *caller, module_metadata);
+  }
+
+  // *** Create the rewriter wrapper helper
+  ILRewriterWrapper reWriterWrapper(&rewriter);
+  reWriterWrapper.SetILPosition(rewriter.GetILList()->m_pNext);
+
+  // *** Modify the Local Var Signature of the method and initialize the new local vars
+  ULONG callTargetStateIndex = static_cast<ULONG>(ULONG_MAX);
+  ULONG exceptionIndex = static_cast<ULONG>(ULONG_MAX);
+  ULONG callTargetReturnIndex = static_cast<ULONG>(ULONG_MAX);
+  ULONG returnValueIndex = static_cast<ULONG>(ULONG_MAX);
+  mdToken callTargetStateToken = mdTokenNil;
+  mdToken exceptionToken = mdTokenNil;
+  mdToken callTargetReturnToken = mdTokenNil;
+  ILInstr* firstInstruction;
+  callTargetTokens->ModifyLocalSigAndInitialize(&reWriterWrapper, caller, 
+      &callTargetStateIndex, &exceptionIndex, 
+      &callTargetReturnIndex, &returnValueIndex, 
+      &callTargetStateToken,
+      &exceptionToken, &callTargetReturnToken, &firstInstruction);
+
+  // ***
+  // BEGIN METHOD PART
+  // ***
+
+  // *** Load instance into the stack (if not static)
+  if (isStatic) {
+    if (caller->type.valueType) {
+      // Static methods in a ValueType can't be instrumented. 
+      // In the future this can be supported by adding a local for the valuetype and initialize it to the default value.
+      // After the signature modification we need to emit the following IL to initialize and load into the stack.
+      //    ldloca.s [localIndex]
+      //    initobj [valueType]
+      //    ldloc.s [localIndex]
+      Warn("*** CallTarget_RewriterCallback(): Static methods in a ValueType cannot be instrumented. ");
+      return E_FAIL;
+    }
+    reWriterWrapper.LoadNull();
+  } else {
+    reWriterWrapper.LoadArgument(0);
+    if (caller->type.valueType) {
+      if (caller->type.type_spec != mdTypeSpecNil) {
+        reWriterWrapper.LoadObj(caller->type.type_spec);
+      } 
+      if (caller->type.name.find("`1"_W) == std::string::npos) {
```

Comment:
I'm not sure I understand the purpose of this check, can you explain it to me? If a type has more than one generic argument, can we exit gracefully?

Reply:
This same pattern is used when writing the BeginMethod calls so any clarification would help me understand that too 😄

Reply:
Ok, yeah, is to handle the case of a generic struct. I did some modifications and refactor to handle this case in a better way with some comments in 40b59b5.

In summary the code changed to:
```cpp
    reWriterWrapper.LoadArgument(0);
    if (caller->type.valueType) {
      if (caller->type.type_spec != mdTypeSpecNil) {
        reWriterWrapper.LoadObj(caller->type.type_spec);
      } 
      else if (!caller->type.isGeneric) {
        reWriterWrapper.LoadObj(caller->type.id);
      } else {
        // Generic struct instrumentation is not supported
        // IMetaDataImport::GetMemberProps and IMetaDataImport::GetMemberRefProps returns 
        // The parent token as mdTypeDef and not as a mdTypeSpec
        // that's because the method definition is stored in the mdTypeDef
        // The problem is that we don't have the exact Spec of that generic
        // We can't emit LoadObj or Box because that would result in an invalid IL.
        // This problem doesn't occur on a class type because we can always relay in the
        // object type.
        return E_FAIL;
      }
    }

```

Reply:
Ah makes sense, thanks!